Automatically strip Authorization header on redirects

dantraMSFT opened the pull request:
Fix #2227

Update Invoke-WebRequest and Invoke-RestMethod cmdlets to strip an Authorization header on redirect.

The FullCLR implementation uses WebRequest to perform the request which silently strips the Authorization header when a redirect occurs. The CoreCLR implementation uses HttpClient to perform the request which does not strip the authorization header. The change explicitly handles the initial redirect, removes the authorization header and submits the request to location in the response.
A new switch, PreserveAuthorizationOnRedirect, disables this handling.

See https://github.com/PowerShell/PowerShell-Docs/pull/1269 for the doc change to document the new switch -PerserveAuthorizationOnRedirect.

dantraMSFT committed:
WebRequestPSCmdlet: Automatically strip Authorization header on redirects

dantraMSFT committed:
WebRequestPSCmdlet: Automatically strip Authorization header on 302 redirects (#1)

WebRequestPSCmdlet: Automatically strip Authorization header on 302 redirects

dantraMSFT committed:
WebRequestPSCmdlet: Automatically strip Authorization header on redirects

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -48,7 +48,23 @@ public HttpResponseException (string message, HttpResponseMessage response) : ba
+        /// gets or sets the PreserveAuthorizationOnRedirect property
+        /// </summary>
+        /// <remarks>
+        /// This property overrides compatability with web requests on Windows.
```
Typo: compatibility

dantraMSFT replied:
Fixed

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -57,7 +73,7 @@ public abstract partial class WebRequestPSCmdlet : PSCmdlet
         internal abstract void ProcessResponse(HttpResponseMessage response);
 
-        #endregion Abstract Methods
+#endregion Abstract Methods
```
Generally, you should not mix formatting changes with functional changes.  Have them as separate PRs to make them more readable and reviewable.

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -367,23 +390,50 @@ internal virtual void FillRequestStream(HttpRequestMessage request)
+                using (client = GetHttpClient(false))
+                using (HttpRequestMessage redirectRequest = GetRequest(response.Headers.Location))
+                {
+                    string msg = string.Format("{0}->{1}", request.RequestUri.AbsoluteUri, response.Headers.Location.AbsoluteUri);
```
You should put this in WebCmdletStrings.resx

dantraMSFT replied:
I'll remove it. Not appropriate for release.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -223,6 +223,43 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
 
+    # Verifies Invoke-WebRequest with -PreserveAuthorizationOnRedirect preserves
+    # the authorization header whe a 302 redirection
+    It "Validate Invoke-WebRequest with -PreserveAuthorizationOnRedirect preserves the authorization header on 302 redirects" {
```
Don't we need to support other 3xx redirect codes?

https://en.wikipedia.org/wiki/URL_redirection#HTTP_status_codes_3xx

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -223,6 +223,43 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
+    # the authorization header whe a 302 redirection
+    It "Validate Invoke-WebRequest with -PreserveAuthorizationOnRedirect preserves the authorization header on 302 redirects" {
+        try {
+            Start-HttpListener -AsJob
```
There's a race condition on Mac where stopping the httplistener and then starting a new one, the OS may not have cleaned up the actual underlying http service so the next start fails as the reservation still exists.  In anycase, you should rebase against upstream/master as this script already starts and stops httplistener for all tests.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -223,6 +223,43 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
+            Start-HttpListener -AsJob
+
+            $headers = @{"Authorization" = "test"}
+            $response = Invoke-WebRequest -Uri "http://localhost:8080/PowerShell?test=redirect" -Headers $headers -PreserveAuthorizationOnRedirect
```
Seems like we should test multiple redirects to make sure it's stripped off the first one and subsequent redirects get handled correctly (3 may be sufficient)

SteveL-MSFT replied:
Should also have a test case where `-PreserveAuthorizationOnRedirect` is specified, but no Authorization header was supplied

PaulHigin commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -367,23 +390,50 @@ internal virtual void FillRequestStream(HttpRequestMessage request)
+                stripAuthorization 
+                )
+            {
+                _cancelToken.Cancel();
```
> _cancelToken.Cancel(); [](start = 15, length = 23)

I am a little confused by this logic.  It looks like an asynchronous call is being cancelled here.  Is there a possibility of a race condition?  Can you add a comment indicating what is being cancelled?

dantraMSFT committed:
Handle all redirection codes

dantraMSFT committed:
Handle all redirection codes

PaulHigin commented:
This seems to have a lot of unrelated changes (26 files now changed) now.

SteveL-MSFT commented:
@dantraMSFT looks like you accidentally pulled in other commits, you need to do a `rebase -i` to clean this up.  Let me know if you need help with this.

dantraMSFT committed:
Refactor tests to use -TestCases to reduce redundant code.

dantraMSFT committed:
Refactor tests to use -TestCases to reduce redundant code.

dantraMSFT committed:
WebRequestPSCmdlet: Automatically strip Authorization header on 302 redirects (#1)

WebRequestPSCmdlet: Automatically strip Authorization header on 302 redirects

dantraMSFT committed:
WebRequestPSCmdlet: Automatically strip Authorization header on 302 redirects (#1)

WebRequestPSCmdlet: Automatically strip Authorization header on 302 redirects

dantraMSFT committed:
Handle all redirection codes

dantraMSFT committed:
Handle all redirection codes

dantraMSFT committed:
Fix redirect tets - move supportedRedirects into script block.

dantraMSFT committed:
Fix redirect tets - move supportedRedirects into script block.

dantraMSFT commented:
fixed

---
In reply to: [306282201](https://github.com/PowerShell/PowerShell/pull/3885#issuecomment-306282201) [](ancestors = 306282201)

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -48,6 +48,22 @@ public HttpResponseException (string message, HttpResponseMessage response) : ba
+        /// authorization header needs to be preserved across redirects.
+        /// </remarks>
+        [Parameter]
+        public virtual SwitchParameter PreserveAuthorizationOnRedirect { get; set; }
```
You need to make the corresponding doc change in PowerShell/PowerShell-Docs and mention it here to link the two

dantraMSFT replied:
It's in progress.  Will submit a PR shortly and update the description to reference it.

SteveL-MSFT replied:
Also, do we need this capability with invoke-restmethod?

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -367,13 +394,77 @@ internal virtual void FillRequestStream(HttpRequestMessage request)
+            (
+                (intCode >= 300 && intCode < 304)
+                ||
+                intCode == 307
```
isn't 308 also valid?

SteveL-MSFT replied:
Since you have this specific checks it seems you should have TestCases that go through the different status codes

dantraMSFT replied:
308 isn't currently supported in CoreCLR - HttpStatusCode doesn't have a define for it so there's no way for the code to see it.

dantraMSFT replied:
An exhaustive -PreserveAuthroizationOnRedirect won't add any value but if you feel strongly enough, I can make these tests cover all redirections as well.

dantraMSFT replied:
Refactored all tests to cover each redirect status code.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -143,6 +142,125 @@ function GetTestData
+        [Microsoft.PowerShell.Commands.WebResponseObject] $Response
+    )
+
+    $index = $Response.RawContent.IndexOf("{")
```
why isn't $response.content | convertfrom-json sufficient?  this code seems unnecessary

dantraMSFT replied:
Because I just want the json part of the raw content. The rest is extraneous for my purposes and it is also not valid json.

SteveL-MSFT replied:
Looking at your usage below, you pass in the output of Invoke-WebRequest to this helper.  This helper isn't necessary.  Line 189 below can just be:

```powershell
$result.Content = $result.Output.Content | ConvertFrom-Json
```

This is assuming the output you create in httplistener is json (and if not, perhaps it should be).

dantraMSFT replied:
The listener does add the output as json and I've even tried setting ContentType and ContentEncoding with no change in behavior. It appears it may be an issue on the client side. For now, I'll update the function to detect byte[] versus  string and convert the byte[] to string before calling ConvertFrom-Json so it will work with arbitrary servers.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -143,6 +142,125 @@ function GetTestData
+    try
+    {
+        $headers = @{"Authorization" = "test"}
+        if ($UsePost)
```
It would be better to have this as a `-testcase` where one of the parameters is $method and you can then pass Get and Post

dantraMSFT replied:
This is only a choice between GET and POST. No other methods are being used.

SteveL-MSFT replied:
lines 182 and 186 below only differ by use of `-Method`.  You should either change `$UsePost` to `$Method` or use splatting so there's only a single Invoke-WebRequest line.

dantraMSFT replied:
I'll update the function to use a ValidateSet and default to GET. I don't want to accept arbitrary methods  since I'm explicitly testing the POST->GET logic on redirects.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -143,6 +142,125 @@ function GetTestData
+    return $result
+}
+
+Describe "Invoke-WebRequest redirect tests" -Tags ("Feature", "Redirect") {
```
Is there a specific reason you want to add a new Describe rather than having your tests part of the existing Invoke-WebRequest tests?

dantraMSFT replied:
I placed the tests in a new describe so I can run them in isolation while developing them.  it turns out appveyor doesn't allow extra tag so I'll be removing it shortly.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -143,6 +142,125 @@ function GetTestData
+        $response.Error | Should BeNullOrEmpty
+
+        # ensure Authorization header has been preserved.
+        $response.Content.Headers -contains "Authorization" | Should Be $true
```
should validate the value as well

dantraMSFT replied:
Will fix.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -143,6 +142,125 @@ function GetTestData
+
+        $response.Error | Should BeNullOrEmpty
+        # ensure Authorization header was stripped
+        $response.Content.Headers -contains "Authorization" | Should Be $true
```
validate the value

dantraMSFT replied:
Will fix

dantraMSFT replied:
Actually, I can't fix it without fixing ConvertTo-Json.   $request.Headers is a WebHeaderCollection which is a NamedValueCollection (ICollection).  ConvertTo-Json (and Select-Object for that matter) treat it as a collection and would have to special case the type.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -143,6 +142,125 @@ function GetTestData
+        $response.Error | Should BeNullOrEmpty
+
+        # ensure user-agent is present (i.e., no false positives )
+        $response.Content.Headers -contains "User-Agent" | Should Be $true
```
why not validate all the headers are there except authorization?

dantraMSFT replied:
I don't know explicitly what other headers 'should' be there so I don't have a source of truth to compare against what is echoed in the response.  I know the authorization header should or should not be in a specific echoed response so I can verify it.

dantraMSFT committed:
Refactor redirect tests to cover all redirect codes for each test.
Remove Redirect tag - not allowed by appveyor.

dantraMSFT committed:
Refactor redirect tests to cover all redirect codes for each test.
Remove Redirect tag - not allowed by appveyor.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -143,6 +142,143 @@ function GetTestData
+    )
+
+    It "Validates Invoke-WebRequest with -PreserveAuthorizationOnRedirect preserves the authorization header on redirect" -TestCases $redirectTests {
+        param($redirectType, $redirectedMethod)
```
Not entirely sure about the Pester syntax, but do you have to have $redirectedMethod here since you don't use it?  Or does Pester complain?

dantraMSFT replied:
$response.Content is a byte array, not string ($result.Output.Content is this byte array).  $response.RawContent has a few lines of text inserted by the underlying API followed by the json form of the original request; inserted by the httplistener.  The purpose of the function is to extract the json from the RawContent and return it explicitly since it is needed to verify the incoming request.

dantraMSFT replied:
I'll see if I can drop the unused $redirectedMethod params.

dantraMSFT replied:
I'm remove the unused parameters.

dantraMSFT committed:
Remove unused $redirectedMethod parameter
Handle string and byte[] in Convert
Update ExecuteRedirectRequest to accept a method name (GET or POST)
Add a delay to Start-HTTPListener to mitigate slower systems.

dantraMSFT committed:
Remove unused $redirectedMethod parameter
Handle string and byte[] in Convert
Update ExecuteRedirectRequest to accept a method name (GET or POST)
Add a delay to Start-HTTPListener to mitigate slower systems.

dantraMSFT committed:
Add Content-Type header in HTTPListener and remove Convert-ResponseContent

dantraMSFT committed:
Add Content-Type header in HTTPListener and remove Convert-ResponseContent

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -143,6 +142,121 @@ function GetTestData
+    return $result
+}
+
+Describe "Invoke-WebRequest redirect tests" -Tags "Feature" {
```
I thought you were going to put these tests into an existing Describe?

dantraMSFT committed:
Add Invoke-RestMethod tests and move tests into existing describe block.
Integrate httpslistener changes

dantraMSFT committed:
Add Invoke-RestMethod tests and move tests into existing describe block.
Integrate httpslistener changes

dantraMSFT commented:
Please squash when merged.

SteveL-MSFT reviewed (approved):
LGTM

lzybkr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -217,6 +240,10 @@ internal virtual HttpRequestMessage GetRequest(Uri uri)
                     else
                     {
+                        if (stripAuthorization && entry.Key == HttpKnownHeaderNames.Authorization.ToString())
+                        {
```
We try to avoid `==` with string comparisons and instead use `string.Equals(a,b, StringComparison.Ordinal)` or the appropriate comparison - this makes it clearer that the author considered culture and case when writing the code.

dantraMSFT replied:
Will fix

lzybkr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -48,6 +48,22 @@ public HttpResponseException (string message, HttpResponseMessage response) : ba
+        /// <remarks>
+        /// This property overrides compatibility with web requests on Windows.
+        /// On FullCLR (WebRequest), authorization headers are stripped during redirect.
+        /// CoreCLR (HTTPClient) does not have this behavior so web reqeusts that work on
```
>reqeusts [](start = 68, length = 8)

Spelling.

dantraMSFT replied:
Will fix.

lzybkr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -410,7 +511,7 @@ protected override void ProcessRecord()
                         }
 
-                        using (HttpRequestMessage request = GetRequest(uri))
+                        using (HttpRequestMessage request = GetRequest(uri, false))
```
>false [](start = 76, length = 5)

Just a suggestion - we like to use named parameters when using bool constants - it makes it much easier to understand the code, so this could be `GetResult(uri, stripAuthorization: false)`.

dantraMSFT replied:
Will change.

lzybkr reviewed (changes requested):
There were a couple of minor issues I'd like to see addressed before I merge.

dantraMSFT committed:
Address review feedback.

dantraMSFT committed:
Change CRLF to LF

dantraMSFT commented:
@lzybkr Would you review my changes so I can get this merged before I go OOF.
Thanks.

lzybkr commented:
I'll squash, but I'll ask you to write the commit message that follows our [guidelines](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#common-engineering-practices) - you can just comment here or squash the changes yourself and update the PR, either way is fine.

dantraMSFT commented:
Description updated.
